code/workspaces/auth-service/src/dataaccess/userRolesRepository.js

```diff
@@ -31,7 +31,12 @@ async function getUser(userId) {
 
 async function getUsers() {
   const allRoles = await UserRoles().find().exec();
-  return allRoles.map(roles => convertToUser(roles));
+  const users = allRoles
+    .filter(roles => roles.userName) // only take users with known user names
+    .map(roles => convertToUser(roles));
+  const uniqueUserIDs = [...new Set(users.map(user => user.userId))];
+  const uniqueUsers = uniqueUserIDs.map(userId => users.filter(user => userId === user.userId)[0]);
+  return uniqueUsers;
```

Comment:
I'm not sure this is the best way of doing this, particularly line 38 as I think that will result in really large number of iterations over the `users` array. The way I would tackle it that should result in fewer iterations would be

```js
const uniqueUsers = users.reduce(
  (currentUniqueUsers, user) => {
    if (!currentUniqueUsers[user.userId]) {
      currentUniqueUsers[user.userId] = user;
    }
    return currentUniqueUsers;
  },
  {},
);
return Object.values(uniqueUsers);
```

I've only typed this freehand in the comment box so might not be spot on, but I think that should provide the same result.

Reply:
Thanks.  I like the Set approach for primitives, but it doesn't work so well for objects, which is what necessitated the extra filter.  I've updated with an approach similar to what you specify.